File: mergeSort/merge_sort.c


#include<stdio.h>
#include<stdlib.h>

int maxInt = 1<<31 - 1;

void printV(int *v,int n){
	int i;
	for(i=0 ; i < n ;i++){
		printf("%d,",v[i]);
	}
}


void merge(int *a,int p,int q,int r){
	int n1 = q - p + 1;
	int n2 = r - q;
	int i,j=0,k=0;


	int *l1 = malloc(sizeof(int)*(n1 +1));
	int *r2 = malloc(sizeof(int)*(n2 +1));
	l1[n1 ] = maxInt;
	r2[n2 ] = maxInt;

	for(i=0; i < n1 ; i++){

		l1[i] = a[p + i];
	}

	for(i=0 ; i < n2;i++){
		r2[i] = a[q + i + 1];
	}

	for(i=p ; i <= r ; i++ ){

		if( l1[j] <= r2[k]){
			a[i] = l1[j++];
		}
		else{
			a[i] = r2[k++];
		}
	}

	free(l1);
	free(r2);
}


void mergeSort(int *a,int p,int r){
	int q;
	if(p < r){
		q = (p + r)/2;
		mergeSort(a,p,q);
		mergeSort(a,q+1,r);
		merge(a,p,q,r);

	}

}



int main(){

	int v[6] = {5,2,4,6,1,3};


	mergeSort(v,0,5);
	printV(v,6);
	return 0;

}

File: mergeSort/Merge_sort.go
package main 

import (
	"fmt"
)

var maxInt int = 1<<31 - 1;

func Merge(a []int,p int, q int, r int){

	var i,j,k int
	n1 := q -p;
	n2 := r-q;
	
	l1 := make([]int,n1 + 1, n1 + 2 )
	r2 := make([]int,n2 ,n2 + 1)
	
	for i,_ = range l1 {
		l1[i] = a[i + p] 

	}
	l1 = l1[:cap(l1)]
	l1[n1 + 1  ] = maxInt
	
	for i,_ = range r2{
		r2[i] = a[i + q + 1]
	}
	r2 = r2[:cap(r2)] 
	r2[n2] = maxInt
	
	
	
	for i = p ; i <= r ; i++{
		
		if l1[j] <= r2[k] {
			//fmt.Print(i,j,k,l1[j],r2[k],"X")
			
			a[i] = l1[j]
			j += 1
		
		} else {
			a[i] = r2[k]
			k += 1
		}
	}
	
}

func MergeSort(a []int,i int,j int){
	
	if i < j {
		
		q := (i+j)/2
		
		MergeSort(a,i,q)
		MergeSort(a,q+1,j)
		
		Merge(a,i,q,j)
	}
}

func main() {

	a := [...]int{5,2,4,6,1,3,8,9}
	b := a[:]
	
	MergeSort(b,0,len(a) - 1)
	
	fmt.Print(a)
}

